tests/virtualization/livemigration/live_migration_src.pm

```diff
@@ -0,0 +1,279 @@
+    if ($vm_type eq "kvm") {
+        $vms->{$vm_name}->{dsturi} = "qemu+ssh://$dst_ip/system";
+        $vms->{$vm_name}->{virsh_cmds} = ["virsh migrate $migrate_opts", "virsh migrate $migrate_opts --tunnelled"];
+    } else {
+        $vms->{$vm_name}->{dsturi} = "xen+ssh://$dst_ip/";
+        $vms->{$vm_name}->{virsh_cmds} = ["virsh migrate --live", "xl migrate"];
+    }
+}
+
+#start a guest, a guest represents a vm for kvm test, two vms for xen test (a fv and a pv)
+sub start_guest {
+    my ($guest, $vms) = @_;
+    if (check_var('SYSTEM_ROLE', 'kvm')) {
+        create_vm($vms, $guest, "kvm", "x86_64");
+    } elsif (check_var('SYSTEM_ROLE', 'xen')) {
+        create_vm($vms, $guest, "fv", "x86_64");
+        create_vm($vms, $guest, "pv", "x86_64");
+    }
+}
+
+# perform live migration test and clean vm up after.
+sub migrate_test {
+    my ($vm_name, $vm_info) = @_;
+    for my $cmd (@{$vm_info->{virsh_cmds}}) {
+        #check if ip file is created. timeouts after 2 mins
+        assert_script_run("check_ip_file(){ for i in \$(seq 1 18);do if [ -f $images_path/guestip/$vm_name ]; then sleep 10; return; else sleep 10; fi; done; false; }; check_ip_file");
+        $vm_info->{ip} = script_output("cat $images_path/guestip/$vm_name");
+
+        if ("$cmd" eq "xl migrate") {
+            record_info("$cmd $vm_name $dst_ip");
+            assert_script_run("$cmd $vm_name $dst_ip");
+        } else {
+            record_info("$cmd $vm_name $vm_info->{dsturi}");
+            assert_script_run("$cmd $vm_name $vm_info->{dsturi}");
+        }
+        #soft_failure bsc#1190309
+        if ($vm_name =~ /fv/ && script_run("ssh root\@$dst_ip ip link show type bridge_slave | grep 'vif.*state DOWN'") == 0) {
+            record_soft_failure("bsc#1190309");
+            assert_script_run("ssh root\@$dst_ip xl list $vm_name");
+        }
+        #visit web server running in vm check it's content
+        validate_script_output("curl -s http://$vm_info->{ip}/guest", sub { m/This is guest $vm_name/ });
+        #test vm disk write and read
+        assert_script_run("echo $vm_name > /tmp/test_vm_write");
+        assert_script_run("curl -s -k -T /tmp/test_vm_write -u bernhard:$testapi::password scp://$vm_info->{ip}/~/test_vm_write");
+        validate_script_output("curl -s -k -u bernhard:$testapi::password scp://$vm_info->{ip}/~/test_vm_write", sub { m/$vm_name/ });
+        assert_script_run("rm $images_path/guestip/$vm_name");
+        #shutdown migrated vm
+        if ($cmd eq "xl migrate") {
+            assert_script_run("ssh root\@$dst_ip xl shutdown $vm_name");
+        } else {
+            assert_script_run("ssh root\@$dst_ip virsh shutdown $vm_name");
+        }
+        #wait for shutdown complete
+        sleep 25;
+        #start again for next migration test
+        assert_script_run("virsh start $vm_name") if ($cmd ne $vm_info->{virsh_cmds}->[-1]);
+    }
+}
+
+#clean up VMs on a host before migration test rerun
+sub cleanup_for_rerun {
```

Comment:
refer to bsc#1190588, just only effect SLE12 SP4 without SLE12 SP5 and 15+ here.
should be:
systemctl("restart libvirtd") if (is_sle('=12-SP4'));

Reply:
I added this line "restart libvirtd" as workaround for that bug.   Later I think that as it's "cleanup" it might be also good to restart libvirtd for any version.  Make sense?

Reply:
no, totally do not agree on this.  if there was no special required, please do not restart libvirtd service included "cleanup" as your mentioned, IMHO.

Reply:
I would vote for Tony's idea -- restarting libvirt service in cleanup_for_rerun totally make sense IMHO.

Reply:
sure, just only refer to bsc#1190588,  "restart libvirtd" as the workaround for SLE12 SP4 as enough.